annif/backend/yake.py

```diff
@@ -0,0 +1,196 @@
+        return self._graph
+
+    def initialize(self):
+        self._initialize_index()
+        self._kw_extractor = yake.KeywordExtractor(
+            lan=self.params['language'],
+            n=self.params['max_ngram_size'],
+            dedupLim=self.params['deduplication_threshold'],
+            dedupFunc=self.params['deduplication_algo'],
+            windowsSize=self.params['window_size'],
+            top=self.params['num_keywords'],
+            features=self.params['features'])
+
+    def _initialize_index(self):
+        if self._index is None:
+            path = os.path.join(self.datadir, self.INDEX_FILE)
+            if os.path.exists(path):
+                self._index = self._load_index(path)
+                self.info(
+                    f'Loaded index from {path} with {len(self._index)} labels')
+            else:
+                self.info('Creating index')
+                self._create_index()
+                self._save_index(path)
+                self.info(f'Created index with {len(self._index)} labels')
+
+    def _save_index(self, path):
+        with open(path, 'w', encoding='utf-8') as indexfile:
+            for label, uris in self._index.items():
+                line = label + '\t' + ' '.join(uris)
+                print(line, file=indexfile)
+
+    def _load_index(self, path):
+        index = dict()
+        with open(path, 'r', encoding='utf-8') as indexfile:
+            for line in indexfile:
+                label, uris = line.strip().split('\t')
+                index[label] = uris.split()
+        return index
+
+    def _create_index(self):
+        index = defaultdict(set)
+        for concept in self.graph.subjects(RDF.type, SKOS.Concept):
+            if (concept, OWL.deprecated, rdflib.Literal(True)) in self.graph:
+                continue
+            uri = str(concept)
+            for label_type in self.label_types:
+                for label in self.graph.objects(concept, label_type):
+                    if not label.language == self.params['language']:
+                        continue
+                    label = self._normalize_label(label)
+                    index[label].add(uri)
+        index.pop('', None)  # Remove possible empty string entry
+        self._index = dict(index)
+
+    def _normalize_label(self, label):
+        label = str(label)
+        if annif.util.boolean(self.params['remove_specifiers']):
+            label = re.sub(r' \(.*\)', '', label)
+        lemmatized_label = self._lemmatize_phrase(label)
+        return self._sort_phrase(lemmatized_label)
+
+    def _lemmatize_phrase(self, phrase):
+        normalized = []
+        for word in phrase.split():
+            normalized.append(
+                self.project.analyzer.normalize_word(word).lower())
+        return ' '.join(normalized)
+
+    def _sort_phrase(self, phrase):
+        words = phrase.split()
+        return ' '.join(sorted(words))
+
+    def _suggest(self, text, params):
```

Comment:
The `params` passed here don't seem to be used for anything else than the `limit` value. Doesn't this mean that the `--backend-param` option cannot be used to set the parameters at runtime? Would it be better to pass `params` around explicitly between the methods and not rely on `self.params` which are determined by the configuration file?

Reply:
The Yake object is now created inside `suggest` and it is instantiated with the explicitly passed `params`; this allows setting those parameters on runtime. However the index is still created or loaded in `initialize`, which doesn't have access to the passed `params`, so switching to a different labels set is not possible on runtime. That would be quite convenient, but I think it would necessitate having separate indices/index files for pref, alt and hiddenLabels.

Reply:
OK - this could perhaps be documented on the wiki page